design/custom-request-timeout.md

```diff
@@ -0,0 +1,362 @@
+# Executive Summary
+
+**Status**: _Draft_
+
+This document describes the design of a new resource in IngressRoute for custom request timeout and retries. 
+
+This new resource will be integrated into Contour 0.11.
+
+# Goals
+
+- Generic resource name for configuring custom request timeouts per IngressRoute
+- Generic resource name for configuring custom retry attempts per IngressRoute and VirtualHosts
+
+# Non-goals
+
+- None
+
+# Background
+
+Contour supports custom request timeout and custom retry attempts via [Ingress Annotations](https://github.com/heptio/contour/blob/master/docs/annotations.md).
+The same functionality has been requested via IngressRoute as well.
+
+# High-level design
+
+At a high level, this document proposes the changes in the IngressRoute's CRD to also support custom request
+timeouts and retries. It covers:
+
+- Fields added for timeout and retries in the YAML file.
+- A high-level coverage on the struct changes required in the code.
+- Arguments passed for Envoy API calls.
+
+## Proposed YAML fields
+
+```yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata:
+    name: github.com
+    namespace: prod
+spec:
+    virtualhost:
+        fqdn: www.github.com
+        # this retryPolicy is applied globally to all routes under this virtual host.
+        # if a retryPolicy is redefined in a route, that policy gets preference for that route only.
+        retryPolicy:
```

Comment:
I just had a look at a retry pattern (src: https://docs.microsoft.com/en-us/azure/architecture/patterns/retry )

> **Cancel.** If the fault indicates that the failure isn't transient or is unlikely to be successful if repeated, the application should cancel the operation and report an exception. For example, an authentication failure caused by providing invalid credentials is not likely to succeed no matter how many times it's attempted.

> **Retry.** If the specific fault reported is unusual or rare, it might have been caused by unusual circumstances such as a network packet becoming corrupted while it was being transmitted. In this case, the application could retry the failing request again immediately because the same failure is unlikely to be repeated and the request will probably be successful.

`4xx` are user based errors and `5xx` are server based. `4xx` if retried, from my understanding, will continue to cause error unless the user corrects it. By complying with the highlighted pattern, in my opinion we shouldn't retry for most of `4xx`. 

However, there are status codes such as [409](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409) and [429](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429) seem retriable.

Regarding your second statement of defining a set of retryPolicies (if I understand this correctly, it is like having custom names for a group of predefined status codes, just like the Envoy API), I fear that it will lead to Envoy [Issue 4103](https://github.com/envoyproxy/envoy/issues/4103). There may come a scenario where users might request new codes to include, eventually leading to numerically accepting status codes.

Reply:
I would prefer not to see anything route or path related added at the virtual host level. The Ingressroute crd is underdeveloped on this point by the virtualhost section is _not_ the catch all for routes. For this iteration, please do not try to implement a per virtualhost default setting. Leave it on the route.